### ISSUE TITLE
Type mismatch between els_protocol and els_transport

### ISSUE THREAD

**Issue by kvelicka** [erlang-ls/erlang_ls#613]
`els_transport` claims that implementation modules should take `binary()` as the 2nd arg to `send`:

```erlang
-callback send(any(), binary())      -> ok.
```

However, `els_server` uses `els_protocol` to format commands and the format is `iolist()`. This can and does cause issues, e.g.:

```
2020-03-25 21:07:44 =CRASH REPORT====
  crasher:
    initial call: els_server:init/1
    pid: <0.161.0>
    registered_name: els_server
    exception error: bad argument: [{io,format,[<0.157.0>,[67,111,110,116,101,110,116,45,76,101,110,103,116,104,58,32,54,54,52,13,10,13,10,<<"{\"id\":0,\"jsonrpc\":\"2.0\",\"result\":{\"capabilities\":{\"codeActio
nProvider\":true,\"codeLensProvider\":{\"resolveProvider\":false},\"completionProvider\":{\"resolveProvider\":false,\"triggerCharacters\":[\":\",\"#\",\"?\",\".\"]},\"definitionProvider\":true,\"documentFormatting
Provider\":true,\"documentHighlightProvider\":true,\"documentRangeFormattingProvider\":false,\"documentSymbolProvider\":true,\"executeCommandProvider\":{\"commands\":[\"4075:replace-lines\",\"4075:server-info\"]},
\"foldingRangeProvider\":true,\"hoverProvider\":true,\"implementationProvider\":true,\"referencesProvider\":true,\"textDocumentSync\":{\"change\":1,\"openClose\":true,\"save\":{\"includeText\":true}},\"workspaceSy
mbolProvider\":true}}}">>],[]],[]},{els_server,handle_request,2,[{file,"/home/karl/code/erlang_ls/src/els_server.erl"},{line,142}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]},{els_server,handle_cast,2,[{file
,"/home/karl/code/erlang_ls/src/els_server.erl"},{line,114}]},{gen_server,try_dispatch,4,[{file,"gen_server.erl"},{line,637}]},{gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,711}]},{proc_lib,init_p_do_app
ly,3,[{file,"proc_lib.erl"},{line,249}]}]
    ancestors: [els_sup,<0.158.0>]
    message_queue_len: 10
    messages: [{'$gen_cast',{request,<<"window/workDoneProgress/create">>,#{token => <<"b7792212-a979-4b25-a17f-746b3fa09c11">>}}},{'$gen_cast',{notification,<<"$/progress">>,#{token => <<"b7792212-a979-4b25-a17f-746b3fa09c11">>,value => #{cancellable => false,kind => begin,message => <<"0 / 53">>,percentage => 0,title => <<"Indexing OTP">>}}}},{'$gen_cast',{request,<<"window/workDoneProgress/create">>,#{token => <<"993075d6-e440-4a92-bf97-a56d32d59821">>}}},{'$gen_cast',{notification,<<"$/progress">>,#{token => <<"993075d6-e440-4a92-bf97-a56d32d59821">>,value => #{cancellable => false,kind => begin,message => <<"0 / 0">>,percentage => 0,title => <<"Indexing Applications">>}}}},{'$gen_cast',{notification,<<"$/progress">>,#{token => <<"993075d6-e440-4a92-bf97-a56d32d59821">>,value => #{kind => end,message => <<"0 / 0">>}}}},{'$gen_cast',{request,<<"window/workDoneProgress/create">>,#{token => <<"abfe74b3-3fb6-40bc-b99c-67ef96337398">>}}},{'$gen_cast',{notification,<<"$/progress">>,#{token => <<"abfe74b3-3fb6-40bc-b99c-67ef96337398">>,value => #{cancellable => false,kind => begin,message => <<"0 / 0">>,percentage => 0,title => <<"Indexing Dependencies">>}}}},{'$gen_cast',{notification,<<"$/progress">>,#{token => <<"abfe74b3-3fb6-40bc-b99c-67ef96337398">>,value => #{kind => end,message => <<"0 / 0">>}}}},{'$gen_cast',{notification,<<"$/progress">>,#{token => <<"b7792212-a979-4b25-a17f-746b3fa09c11">>,value => #{cancellable => false,kind => report,message => <<"0 / 53">>,percentage => 0}}}},{'$gen_cast',{notification,<<"$/progress">>,#{token => <<"b7792212-a979-4b25-a17f-746b3fa09c11">>,value => #{cancellable => false,kind => report,message => <<"1 / 53">>,percentage => 1}}}}]
    links: [<0.159.0>]
    dictionary: []
    trap_exit: false
    status: running
    heap_size: 4185
    stack_size: 27
    reductions: 1610929
  neighbours:
```

The problem is that the list passed to `els_stdio` (this isn't obvious from the stack trace) contains a list _and_ a binary, which `io:format` appears to dislike.

There's numerous and trivial fixes to this but I'm not sure what direction you'd like to take with the protocol - once clarified I can send a patch. Personally, I'd go with formatting data to binaries at `els_protocol` level as I've found working with iolists unreliable - this bug is a case in point.

The bug is definitely in `0.3.0` but I'm pretty sure it's on `master` too.

**Comment by robertoaloi** [comment]
And yes, this was a regression for an undocumented feature in OTP 21.0 and fixed in OTP 21.1:

`iodata()` has always been allowed by `io:format/X` but not documented.

References:

http://erlang.2086793.n4.nabble.com/io-format-1-2-in-OTP-21-doesn-t-take-iodat-anymore-td4724022.html
https://github.com/erlang/otp/commit/2d04d2bdf95c3a5f143ed2a5b9295fbf197b9b0b

**Comment by kvelicka** [comment]
Aha, interesting find!
